### ISSUE TITLE
New Rule: no-explicit-coercion

### ISSUE THREAD

**Issue by kjetilhartveit** [eslint/eslint#19517]
### Rule details

Disallow explicit type conversions

### Related ECMAScript feature

Core types

### What type of rule is this?

Warns about a potential problem

### Example code

```js
const b = +foo;
const b1 = Number(foo);
const n = String(foo);
const sorted = myArray.toSorted((a, b) => +a - +b);
const sorted2 = myArray.toSorted((a, b) => Number(a) - Number(b));
```

### Why should this rule be in the core instead of a plugin?

Type conversions, both explicit and implicit, is an escape-hatch from type checking. Therefore there should be safety measures in place to prevent this from happening unintentionally.

I often find myself writing code which converts dates to numbers when sorting. An example is `const sorted = myArray.toSorted((a, b) => +a - +b)`. The problem is, if `myArray` changes in the future, e.g. from dates to objects, then TypeScript won't notice it. This is a realistic source for bugs in code which is quite possibly not noticed until at runtime (and maybe after some digging).

Yes, there exists a rule called https://eslint.org/docs/latest/rules/no-implicit-coercion which can warn you when you try to use `+foo` , but then you would just change your code to use `Number(foo)` and the same risks are still there. `Number()` accepts any value.

I propose that there exists a rule (may or may not be recommended/turned on by default) that warns you when you try to convert types, even when explicit. This would make a developer think twice before using `+a` or `Number(a)` in a sort function (or elsewhere), and hopefully opt to use other methods, like `a.getTime()` instead. 

I'm not sure what the best course of action would be regarding this rule (no-explicit-coercion) co-existing alongside no-implicit-coercion. Would someone who would like to opt out of both implict and explicit type coercions enable both rules, or should there rather be one rule for both?

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

I could possibly submit a pull request, but just haven't done it before so a bit uncertain. I assume one could take some inspiration from the no-implicit-coercion rule.

**Comment by nzakas** [comment]
Thanks for the suggestion. I'm not sure this is a universally accepted principle about type coercions, and as such, isn't a good candidate for inclusion in the core.

You can definitely create a custom rule for this, however. Most LLMs are pretty good at creating custom ESLint rules, so I'd suggest starting there if you're unfamiliar with how to write rules.

**Comment by kjetilhartveit** [comment]
Ok, I see. I can try to create a custom rule, try it out and take it from there.

**Comment by surjikal** [comment]
@kjetilhartveit I think you're 100% correct in your issue description, it's a gap in eslint no doubt IMO.

Unfortunately, anyone that views the compiler as the enemy, inevitably use something like `String` or `Number` to defeat it.

```typescript
declare const x: undefined | symbol | number | string | Date;
const a = x as string;            // error: @typescript-eslint/consistent-type-assertions
const b = x as unknown as string; // error: @typescript-eslint/consistent-type-assertions
const c = `${x}`;                 // error: @typescript-eslint/restrict-template-expressions
const d = x.toString();           // error: (tsc) 'x' is possibly 'undefined'
const e = String(x);              // no error!
```

Am looking to warn on `String`, `Number`, and `Boolean` to steer people towards glorious strictness and type safety.

Best I've found so far is disabling it via `no-restricted-syntax` using selector selector `CallExpression[callee.name=/^(Boolean|Number|String)$/]`.
